### PR TITLE
Common code for loading, caching, and error handling

### PR DESCRIPTION
**Pull request by nategbrown9**
The bulk of this PR is the creation of an AbstractResoruceRepo to allow reuse of the loading, caching, and error handling of Bridge calls.  This should allows us to add new endpoints with with offline caching with minimal new code. As an example, this PR includes adding an AppConfigRepo.

### CODE REVIEW

File: shared/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/apis/AuthenticationApi.kt [hunk 1]
```diff
@@ -10,12 +10,11 @@ import io.ktor.client.features.logging.*
 import org.sagebionetworks.bridge.kmm.shared.models.SignIn
 import org.sagebionetworks.bridge.kmm.shared.models.UserSessionInfo
 
-internal class AuthenticationApi(basePath: kotlin.String = "https://webservices.sagebridge.org") : AbstractApi() {
-    override val _basePath = basePath
+internal class AuthenticationApi(basePath: kotlin.String = "https://webservices.sagebridge.org") : AbstractApi(basePath) {
```
Comment (syoung-smallwisdom) [inline shared/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/apis/AuthenticationApi.kt]: BASE_PATH

File: shared/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/apis/ConsentApi.kt [hunk 2]
```diff
@@ -4,10 +4,9 @@ import io.ktor.client.*
 import org.sagebionetworks.bridge.kmm.shared.models.ConsentSignature
 import org.sagebionetworks.bridge.kmm.shared.models.UserSessionInfo
 
-internal class ConsentApi(basePath: kotlin.String = "https://webservices.sagebridge.org", httpClient: HttpClient): AbstractApi() {
+internal class ConsentApi(basePath: kotlin.String = "https://webservices.sagebridge.org", httpClient: HttpClient): AbstractApi(basePath) {
```
Comment (syoung-smallwisdom) [inline shared/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/apis/ConsentApi.kt]: BASE_PATH

File: shared/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/cache/ResourceDatabaseHelper.kt [hunk 3]
```diff
@@ -1,14 +1,16 @@
 package org.sagebionetworks.bridge.kmm.shared.cache
 
 import com.squareup.sqldelight.EnumColumnAdapter
-import com.squareup.sqldelight.runtime.coroutines.asFlow
-import com.squareup.sqldelight.runtime.coroutines.mapToOneOrNull
+//import com.squareup.sqldelight.runtime.coroutines.asFlow
+//import com.squareup.sqldelight.runtime.coroutines.mapToOneOrNull
```
Comment (syoung-smallwisdom) [inline shared/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/cache/ResourceDatabaseHelper.kt:5]: Remove commented out code

File: shared/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/cache/ResourceType.kt [hunk 4]
```diff
@@ -1,17 +1,33 @@
...
 }
 
-fun <T: Any> Resource.loadResource(classOfT: KClass<T>): T {
-    val jsonCoder = Json {  }
-    return jsonCoder.decodeFromString(PolymorphicSerializer(classOfT), json)
+enum class ResourceStatus {
+    SUCCESS(), //Successfully loaded from Bridge
+    PENDING(), //Remote web call is in process
+    RETRY(),   //Remote web call failed due to a transient error and could be retried
+    FAILED()   //Remote web call failed
```
Comment (syoung-smallwisdom) [inline shared/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/cache/ResourceType.kt]: Why empty parenthesis?

File: shared/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/cache/ResourceType.kt [hunk 5]
```diff
@@ -1,17 +1,33 @@
...
+}
+
+sealed class ResourceResult<out T : Any> {
+    data class Success<out T : Any>(val data: T, val status: ResourceStatus) : ResourceResult<T>()
+    //TODO: Failed reason
```
Comment (syoung-smallwisdom) [inline shared/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/cache/ResourceType.kt]: sign and date all TODO comments

File: shared/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/repo/AbstractResourceRepo.kt [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+                        val resourceJson = remoteLoad(identifier)
+                        val resource = Resource(
+                            identifier,
+                            resourceType,
+                            resourceJson,
+                            Clock.System.now().toEpochMilliseconds(),
+                            ResourceStatus.SUCCESS
+                        )
+                        database.insertUpdateResource(resource)
+                    } catch (t: Throwable) {
```
Comment (syoung-smallwisdom) [inline shared/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/repo/AbstractResourceRepo.kt]: Nit: one letter ivars are hard to read and search on. How about `err`?

File: shared/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/repo/AbstractResourceRepo.kt [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+                            status
+                        )
+                        database.insertUpdateResource(resource)
+
+                    }
+
+                }
+
+
+                //Need to make sure webcall is not already in flight
```
Comment (syoung-smallwisdom) [inline shared/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/repo/AbstractResourceRepo.kt]: Is this a TODO? there's no function call after it.

File: shared/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/repo/AbstractResourceRepo.kt [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+    }
+
+
+
+    inline fun <reified T: Any>processResult(resource: Resource?) : ResourceResult<T> {
+        resource?.let {
+
+            val model = it.loadResource<T>()
+            if (model != null) {
+                //If we have a model object, return it as a success
```
Comment (syoung-smallwisdom) [inline shared/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/repo/AbstractResourceRepo.kt]: Nit: Erin M. likes periods at the ends of sentences. Now I can't unsee them. ;)

File: shared/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/repo/AbstractResourceRepo.kt [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+                ResourceStatus.PENDING -> {
+                    return ResourceResult.InProgress
+                }
+                else -> {
+                    return ResourceResult.Failed(it.status)
+                }
+
+            }
+
+        } ?: return ResourceResult.InProgress //If we don't have a resource, it hasn't been downloaded yet
```
Comment (syoung-smallwisdom) [inline shared/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/repo/AbstractResourceRepo.kt]:
```suggestion
        } ?: return ResourceResult.InProgress // If we don't have a resource, it hasn't been downloaded yet.
```

File: shared/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/repo/AbstractResourceRepo.kt (hunk 8 above)
Comment (syoung-smallwisdom) [inline shared/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/repo/AbstractResourceRepo.kt]:
```suggestion
                // If we have a model object, return it as a success.
```

File: shared/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/repo/AbstractResourceRepo.kt [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+                        var status = ResourceStatus.FAILED
+                        var lastUpdate = curResource?.lastUpdate ?: Clock.System.now().toEpochMilliseconds()
+                        var json = curResource?.json
+                        when (t) {
+
+                            is ResponseException -> {
+                                when(t.response.status) {
+                                    HttpStatusCode.NotModified -> {
+                                        // 304 not modified
+                                        //Updated last modified time
```
Comment (syoung-smallwisdom) [inline shared/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/repo/AbstractResourceRepo.kt]:
```suggestion
                                        // Updated last modified time.
```

File: shared/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/repo/AbstractResourceRepo.kt [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+                                        status = ResourceStatus.SUCCESS
+                                        lastUpdate = Clock.System.now().toEpochMilliseconds()
+                                    }
+
+                                    HttpStatusCode.Unauthorized -> {
+                                        // 401 unauthorized
+                                        // User hasn't authenticated or re-auth has failed
```
Comment (syoung-smallwisdom) [inline shared/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/repo/AbstractResourceRepo.kt]:
```suggestion
                                        // User hasn't authenticated or re-auth has failed.
```

**Review by syoung-smallwisdom — APPROVED** [review]
Bunch of nits and some questions.

**Review by liujoshua — APPROVED** [review]

**Pull request merged**